docs/design/clusterVsNamespaceScoped.md

```diff
@@ -0,0 +1,491 @@
+contained in the base struct `ReconcilerCommon`.
+Then there is a \*Program struct, which includes in each \*Program’s Program struct and the base
+struct `ReconcilerCommon`.
+Below are the bpfman-agent structures, but the bpfman-operator follows the same pattern.
+
+```go
+type ReconcilerCommon struct {
+	client.Client
+	Scheme       *runtime.Scheme
+	GrpcConn     *grpc.ClientConn
+	BpfmanClient gobpfman.BpfmanClient
+	Logger       logr.Logger
+	NodeName     string
+	progId       *uint32
+	finalizer    string
+	recType      string
+	appOwner     metav1.Object // Set if the owner is an application
+}
+
+type FentryProgramReconciler struct {
+	ReconcilerCommon
+	currentFentryProgram *bpfmaniov1alpha1.FentryProgram
+	ourNode              *v1.Node
+}
+
+type FexitProgramReconciler struct {
+	ReconcilerCommon
+	currentFexitProgram *bpfmaniov1alpha1.FexitProgram
+	ourNode             *v1.Node
+}
+
+:
+```
+
+### interface
+
+The `bpfmanReconciler` interface defines the set of methods the \*Program structs must implement to use the
+common reconciler code. 
+Below are the bpfman-agent structures, but the bpfman-operator uses a `ProgramReconciler`, which follows
+the same pattern.
+
+```go
+type bpfmanReconciler interface {
+	SetupWithManager(mgr ctrl.Manager) error
+	Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error)
+	getFinalizer() string
+	getOwner() metav1.Object
+	getRecType() string
+	getProgType() internal.ProgramType
+	getName() string
+	getExpectedBpfPrograms(ctx context.Context)
+    (*bpfmaniov1alpha1.BpfProgramList, error)
+	getLoadRequest(bpfProgram *bpfmaniov1alpha1.BpfProgram,
+    mapOwnerId *uint32) (*gobpfman.LoadRequest, error)
+	getNode() *v1.Node
+	getBpfProgramCommon() *bpfmaniov1alpha1.BpfProgramCommon
+	setCurrentProgram(program client.Object) error
+	getNodeSelector() *metav1.LabelSelector
+	getBpfGlobalData() map[string][]byte
+	getAppProgramId() string
+}
+```
+
+There are also some common reconciler functions that perform common code.
+
+```go
+func (r *ReconcilerCommon) reconcileCommon(ctx context.Context,
+  rec bpfmanReconciler,
+	programs []client.Object) (bool, ctrl.Result, error) {
+:
+}
+
+func (r *ReconcilerCommon) reconcileBpfProgram(ctx context.Context,
+	rec bpfmanReconciler,
+	loadedBpfPrograms map[string]*gobpfman.ListResponse_ListResult,
+	bpfProgram *bpfmaniov1alpha1.BpfProgram,
+	isNodeSelected bool,
+	isBeingDeleted bool,
+	mapOwnerStatus *MapOwnerParamStatus)
+(bpfmaniov1alpha1.BpfProgramConditionType, error) {
+:
+}
+
+func (r *ReconcilerCommon) reconcileBpfProgramSuccessCondition(
+	isLoaded bool,
+	shouldBeLoaded bool,
+	isNodeSelected bool,
+	isBeingDeleted bool,
+	noContainersOnNode bool,
+  mapOwnerStatus *MapOwnerParamStatus) bpfmaniov1alpha1.BpfProgramConditionType {
+:
+}
+```
+
+So looks something like this:
+
+```
+                     --- FentryProgramReconciler
+                     |     func (r *FentryProgramReconciler) getFinalizer() string {}
```

Comment:
```suggestion
While the contents are the same for the namespace and cluster-scoped CRD in most cases, Kubernetes requires different 
CRD for each type.
```

Reply:
Done (replaced `To add Namespace scoped BPF Program and \*Program CRDs, Kubernetes requires new CRDs for each type.` with your text. Did add the word `mostly` the same, since we had to remove namespace from ContainerSelector.)